File: leetcode/problems/problem-specific/17.letter-combinations-of-a-phone-number/17.letter-combinations-of-a-phone-number.go
package main

func letterCombinations(digits string) []string {
	if len(digits) <= 0 {
		return []string{}
	}

	MAPPINGS := []string{"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	res := []string{}
	letterCombinationsHelper(digits, MAPPINGS, &res, "", 0)
	return res
}

func letterCombinationsHelper(digits string, MAPPINGS []string, res *[]string, combination string, index int) {
	if index == len(digits) {
		*res = append(*res, combination)
		return
	}

	letters := MAPPINGS[digits[index]-'0']
	for _, v := range letters {
		letterCombinationsHelper(digits, MAPPINGS, res, combination+string(v), index+1)
	}
}

File: leetcode/problems/problem-specific/17.letter-combinations-of-a-phone-number/17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python3
#
# [17] Letter Combinations of a Phone Number
#

# @lc code=start

# * Backtracking Solution | O(4^n) Time | O(n) Space
# * n -> The length of digits

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        LETTERS = [
            '0', '1', "abc", "def", "ghi",
            "jkl", "mno", "pqrs", "tuv", "wxyz"
        ]
        res = []
        if not digits:
            return res

        def letterCombinationsHelper(idx=0, combination=[]):
            if len(combination) == len(digits):
                res.append(''.join(combination))
                return

            for c in LETTERS[ord(digits[idx]) - ord('0')]:
                combination.append(c)
                letterCombinationsHelper(idx + 1, combination)
                combination.pop()

        letterCombinationsHelper()
        return res

# @lc code=end
